### PR TITLE
Add MultiHeadAttention Layer decomposition for PyTorch

### PR DESCRIPTION
**Pull request by lapid92**

### CODE REVIEW

File: model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 1]
```diff
@@ -0,0 +1,718 @@
...
+        Matches MultiHeadAttention node.
+        """
+        super().__init__(matcher_instance=NodeOperationMatcher(nn.MultiheadAttention))
+
+    @staticmethod
+    def _get_weight_by_name(mha_node, w_str):
+        return [k for k in mha_node.weights.keys() if w_str in k][0]
+
+    def _project_input(self, graph, mha_node, params):
+        # (B, q_seq, q_dim*n_h) --> (B, q_dim*n_h, q_seq) --> (B, q_dim*n_h, q_seq)
```
Comment (elad-c) [inline model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py:121]: Add comment to explain the comment

File: model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 2]
```diff
@@ -0,0 +1,718 @@
...
+        # (B, kv_seq, k_dim) --> (B, k_dim, kv_seq) --> (B, q_dim*n_h, kv_seq)
+        # (B, kv_seq, v_dim) --> (B, v_dim, kv_seq) --> (B, q_dim*n_h, kv_seq)
+
+        factor = (params.qdim ** -0.5)
```
Comment (elad-c) [inline model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py:125]: Add comment  to explain the weight calcualtion process

File: model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 3]
```diff
@@ -0,0 +1,718 @@
...
+            in_bias = False
+            q_node_weights = {KERNEL: qk}
+            k_node_weights = {KERNEL: kk}
+            v_node_weights = {KERNEL: vk}
+
+        # project query, key, value
+
+        query_name = f'{mha_node.name}_query_in_proj'
+
+        # (B, q_seq, q_dim*n_h) --> (B, q_dim*n_h, q_seq)
```
Comment (elad-c) [inline model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py:152]: Add comment explaining why the transepose is needed

File: model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 4]
```diff
@@ -0,0 +1,718 @@
...
+                          input_shape=params.v_transpose_in_shape,
+                          output_shape=params.v_proj_in_shape,
+                          weights=v_node_weights,
+                          layer_class=nn.Conv1d)
+        graph.add_node_with_in_edges(v_node, [v_transpose_node])
+
+        return q_transpose_node, k_transpose_node, v_transpose_node, q_node, k_node, v_node
+
+    @staticmethod
+    def _arrange_before_slice(graph, mha_node, q_node, k_node, v_node, params):
```
Comment (elad-c) [inline model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py:223]: change to "split"

File: model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 5]
```diff
@@ -0,0 +1,718 @@
...
+                                      layer_class=torch.split,
+                                      op_call_args=[1],
+                                      op_call_kwargs={DIM: 1},
+                                      functional_op=torch.split)
+        graph.add_node_with_in_edges(v_slice_node, [v_transpose_node])
+
+        return q_slice_node, k_slice_node, v_slice_node
+
+        #
+        # # (B, n_h, q_seq, q_dim) --> (B, 1, q_seq, q_dim) --> (B, q_seq, q_dim)
```
Comment (elad-c) [inline model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py:342]: delete unnecessary comment

File: model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 6]
```diff
@@ -0,0 +1,718 @@
...
+                                      op_call_args=[],
+                                      op_call_kwargs={},
+                                      functional_op=torch.matmul)
+        graph.add_node_with_in_edges(matmulv_node, [softmax_node, get_v_node])
+
+        return matmulv_node
+
+    @staticmethod
+    def _stack_heads_reshape(graph, name, att_head_output_nodes, params):
+        # [(B, 1, q_seq, q_dim)]  * n_h --> (B, 1, n_h, q_seq, q_dim) --> (B, n_h, q_seq, q_dim)
```
Comment (elad-c) [inline model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py:513]: change to concat on 2nd axis instead of stack&reshape

File: model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 7]
```diff
@@ -0,0 +1,718 @@
...
+            if mha_node.framework_attr[ADD_ZERO_ATTN] is not False:
+                raise Exception('Add Zero Attn feature is Not Implemented')
+
+        # Add Bias KV feature is Not Implemented
+        if ADD_BIAS_KV in mha_node.framework_attr.keys():
+            if mha_node.framework_attr[ADD_BIAS_KV] is not False:
+                raise Exception('Add Bias KV feature is Not Implemented')
+
+        # Dropout is Not Implemented
+        if DROPOUT in mha_node.framework_attr:
```
Comment (elad-c) [inline model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py:671]: no need for this limitation: we only run inference mode, so dropout is an identity

File: model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 8]
```diff
@@ -0,0 +1,718 @@
...
+
+        # Only batch first network is supported
+        if BATCH_FIRST in mha_node.framework_attr.keys():
+            if mha_node.framework_attr[BATCH_FIRST] is not True:
+                raise Exception('Only batch first network is supported, please fix your network')
+        else:
+            raise Exception('Only batch first network is supported, please fix your network')
+
+        # Add Zero Attn feature is Not Implemented
+        if ADD_ZERO_ATTN in mha_node.framework_attr.keys():
```
Comment (elad-c) [inline model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py:661]: Check if we need to support this

File: model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 9]
```diff
@@ -0,0 +1,718 @@
...
+            if mha_node.framework_attr[ADD_ZERO_ATTN] is not False:
+                raise Exception('Add Zero Attn feature is Not Implemented')
+
+        # Add Bias KV feature is Not Implemented
+        if ADD_BIAS_KV in mha_node.framework_attr.keys():
+            if mha_node.framework_attr[ADD_BIAS_KV] is not False:
```
Comment (elad-c) [inline model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py:667]: Check if we need to support this

**Review by elad-c — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 10]
```diff
@@ -0,0 +1,718 @@
...
+
+    def __init__(self):
+        """
+        Matches MultiHeadAttention node.
+        """
+        super().__init__(matcher_instance=NodeOperationMatcher(nn.MultiheadAttention))
+
+    @staticmethod
+    def _get_weight_by_name(mha_node, w_str):
+        return [k for k in mha_node.weights.keys() if w_str in k][0]
```
Comment (reuvenperetz) [inline model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py:118]: Why not use mha node's get_weights_by_keys method?

File: model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 11]
```diff
@@ -0,0 +1,718 @@
...
+                                           layer_class=torch.reshape,
+                                           op_call_args=[params.attn_reshape_shape],
+                                           op_call_kwargs={},
+                                           functional_op=torch.reshape)
+        graph.add_node_with_in_edges(attn_reshape_node, [transpose_node])
+
+        return attn_reshape_node
+
+    def _project_output(self, graph, mha_node, attn_reshape_node, params):
+
```
Comment (reuvenperetz) [inline model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py:567]: Please add comments and type hints.

File: model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 12]
```diff
@@ -0,0 +1,718 @@
...
+        Args:
+            graph: input graph
+            mha_node: MHA node to substitute inputs and outputs with
+            q_permute_node: 1st input to MHA node
+            k_permute_node: 2nd input to MHA node
+            v_permute_node: 3rd input to MHA node
+            output_permute_node: output node of MHA node
+
+        Returns:
+
```
Comment (reuvenperetz) [inline model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py:632]: Please remove the "returns" line

File: model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 13]
```diff
@@ -0,0 +1,718 @@
...
+                                            op_call_args=[1, 2],
+                                            op_call_kwargs={},
+                                            functional_op=torch.transpose)
+        graph.add_node_with_in_edges(transpose_out_node, [proj_out_node])
+        return transpose_out_node
+
+    @staticmethod
+    def _connect_to_graph(graph, mha_node, q_node, k_node, v_node, output_permute_node):
+        """
+        connect subgraph to input graph
```
Comment (reuvenperetz) [inline model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py:622]: Please add type hints

File: model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 14]
```diff
@@ -0,0 +1,718 @@
...
+
+        # check for MHA Parameters
+
+        # Only batch first network is supported
+        if BATCH_FIRST in mha_node.framework_attr.keys():
+            if mha_node.framework_attr[BATCH_FIRST] is not True:
+                raise Exception('Only batch first network is supported, please fix your network')
+        else:
+            raise Exception('Only batch first network is supported, please fix your network')
+
```
Comment (reuvenperetz) [inline model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py:659]: Please let's use a different error message

**Review by reuvenperetz — CHANGES_REQUESTED** [review]

**Review by elad-c — APPROVED** [review]
